### ISSUE TITLE
Associative arrays in tests are poorly supported

### ISSUE THREAD

**Issue by anowlcalledjosh** [koalaman/shellcheck#870]
- Rule Id: SC2144, SC1073
- My shellcheck version: online
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit
- [ ] It's not reproducible on shellcheck.net, but I think that's because it's an OS, configuration or encoding issue

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash

declare -A arr

[[ -v arr[$index] ]] && echo "${arr[$index]}"
```

#### Here's what shellcheck currently says:

```
Line 5:
[[ -v arr[$index] ]] && echo "${arr[$index]}"
      ^-- SC2144: -v doesn't work with globs. Use a for loop.
```

#### Here's what I wanted or expected to see:

Shellcheck shouldn't complain about this, as far as I can tell. It looks like this is [valid syntax](http://stackoverflow.com/a/26931860/5951320) since bash 4.2 ([possibly 4.3?](http://stackoverflow.com/a/30356161/5951320)).

It gets even more confused if you have the index in quotes (necessary if the index could be a space, otherwise you get weird errors like `bash: [: arr[: unary operator expected`):

```
Line 5:
[[ -v arr["$index"] ]] && echo "${arr[$index]}"
^-- SC1073: Couldn't parse this test expression.
      ^-- SC1019: Expected this to be an argument to the unary condition.
                   ^-- SC1020: You need a space before the ]].
                   ^-- SC1072: Missing space before ]. Fix any mentioned problems and try again.
```

**Comment by anowlcalledjosh** [comment]
Actually, it looks like `test -v` is not supported at all:

```
[ -v foo ]
     ^-- SC2154: foo is referenced but not assigned.
```

this is surely exactly the usecase for `-v`...

**Comment by koalaman** [comment]
You're right of course. This should be significantly improved in 41b6e3d.

This doesn't fix the parsing for `[[ -v var["$index"] ]]` , but this isn't correct anyways. You should single quote the expression: `[[ -v 'var[$index]' ]]`. You can see the difference when `index="]"`

**Comment by anowlcalledjosh** [comment]
Argh, I can never get the quoting right. Thanks!
